[FSDP] Add training_state parameter to fsdp_pre_all_gather hook

yang-yu-hang opened the pull request:
Summary:
Add support for passing training state information to FSDP tensor subclass hooks during all-gather operations. This enables tensor subclasses to differentiate between forward and backward passes for state-aware optimizations like conditional quantization.

**Changes:**

1. **Hook Signature Extension** (`_fsdp_param.py`, `_fsdp_param_group.py`):
   - Added `param_group` parameter to `FSDPParam.__init__()` to maintain reference to parent `FSDPParamGroup`
   - Extended hook signature detection to support new 2-parameter signature: `fsdp_pre_all_gather(mesh, training_state)`
   - Maintained backward compatibility with existing 1-param and 5-param signatures
   - Pass `TrainingState` from parent group to tensor subclass hooks, defaulting to `TrainingState.IDLE` when `param_group` is None

2. **Implementation Details:**
   - Used `TYPE_CHECKING` guard to avoid circular import with `FSDPParamGroup`
   - Reference cycle between `FSDPParam` and `FSDPParamGroup` is safe (parent-child have same lifetime)
   - Added `TrainingState` import from `_fsdp_common`

**Benefits:**
- Enables tensor subclasses to implement state-aware behavior (e.g., different quantization strategies for forward vs backward)
- Maintains full backward compatibility with existing hooks
- Clean API extension without breaking changes

Differential Revision: D86938522




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:
```diff
@@ -147,6 +147,7 @@ def __init__(
                 shard_placement_fn,
                 mp_policy,
                 offload_policy,
+                param_group=self,  # Ref cycle is OK: parent-child have same lifetime
```
is it enough to pass in training_state=self.training_state?

FSDPParamGroup and FSDPParam pointing each other worries me

weifengpy replied:
oops. do we have to update _fsdp_state.py too? 

https://github.com/pytorch/pytorch/blob/c56655268b4ae575ee4c89c312fd93ca2f5b3ba9/torch/distributed/fsdp/_fully_shard/_fsdp_state.py#L241-L268

actually passing param_group might be less verbose. do you see any cycle ref errors? If no, we can do it

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_param.py:
```diff
@@ -730,22 +737,39 @@ def all_gather_inputs(self) -> list[torch.Tensor]:  # 1D
+                    # New signature with training_state
+                    training_state = (
+                        self._param_group._training_state
+                        if self._param_group is not None
```
is there a real case that self._param_group is None?

yang-yu-hang committed:
[FSDP] Add training_state parameter to fsdp_pre_all_gather hook (#167886)

Summary:

Add support for passing training state information to FSDP tensor subclass hooks during all-gather operations. This enables tensor subclasses to differentiate between forward and backward passes for state-aware optimizations like conditional quantization.

**Changes:**

1. **Hook Signature Extension** (`_fsdp_param.py`, `_fsdp_param_group.py`):
   - Added `param_group` parameter to `FSDPParam.__init__()` to maintain reference to parent `FSDPParamGroup`
   - Extended hook signature detection to support new 2-parameter signature: `fsdp_pre_all_gather(mesh, training_state)`
   - Maintained backward compatibility with existing 1-param and 5-param signatures
   - Pass `TrainingState` from parent group to tensor subclass hooks, defaulting to `TrainingState.IDLE` when `param_group` is None

2. **Implementation Details:**
   - Used `TYPE_CHECKING` guard to avoid circular import with `FSDPParamGroup`
   - Reference cycle between `FSDPParam` and `FSDPParamGroup` is safe (parent-child have same lifetime)
   - Added `TrainingState` import from `_fsdp_common`

**Benefits:**
- Enables tensor subclasses to implement state-aware behavior (e.g., different quantization strategies for forward vs backward)
- Maintains full backward compatibility with existing hooks
- Clean API extension without breaking changes

Differential Revision: D86938522
